### DISCUSSION TITLE
filter map objects using javascript

### DISCUSSION THREAD

**Discussion by NitishKumar525** [practical-tutorials/project-based-learning#621] [Q&A] [answered]
I have a map object with key value pair as shown below

```
    [
      0,
      {
        testErrorSummary: "ServiceC\nV4/\n\tcom.dmv.smp",
        appErrorSummary: '{"handler":null,"sourcetype":null}',
        invocations: [
          { radar: "", testCaseMilestone: "1G05" },
          { radar: "3450", testCaseMilestone: "1G15" },
        ],
      },
    ],
    [
      1,
      {
        testErrorSummary: 'java: \nEx:\n <"()">\n <" "code": "403"">',
        appErrorSummary: '{"handler":null,"sourcetype":null}',
        invocations: [
          { radar: "1532", testCaseMilestone: "1G12" },
          { radar: "1954", testCaseMilestone: "1G24" },
        ],
      },
    ],
    [
      2,
      {
        testErrorSummary: "com.dmv: Timeout while waiting.",
        appErrorSummaryKeyword: 'Cannot invoke "com.dmv.getEcid()"',
        hasRadarHistory: false,
        invocations: [
          { radar: "3444", testCaseMilestone: "1G45" },
          { radar: "", testCaseMilestone: "1G90" },
        ],
      },
    ],
  ]);
```

I have a search input where i enter the string. Based on this string value, i want to filter this above map object. Suppose user enters "50". So i need to check for properties `testErrorSummary`, `appErrorSummary` and `radar` and return only those mapped objects which matches with user entries.

So in this case, user enters 50, i am expecting the below result because those get matched with `radar` having 50 in it. So `filterFailureInput = 50`

**Expected output-**

```
 new Map([
    [
      0,
      {
        testErrorSummary: "ServiceC\nV4/\n\tcom.dmv.smp",
        appErrorSummary: '{"handler":null,"sourcetype":null}',
        invocations: [
          { radar: "", testCaseMilestone: "1G05" },
          { radar: "3450", testCaseMilestone: "1G15" },
        ],
      },
    ],
  ]);
```

In order to achieve this way of filtering, i am doing the following but i dont get the expected filter results. Can someone let me know where i am going wrong with this.

```
 const searchException = () => {
        return new Map(
          Array.from(input).filter(([key, value]) => {
            const matchInvocations = value.invocations.some(
              (invocation) => {
                (invocation?.radar &&
                  invocation?.radar.includes(filterFailureInput)) ||
                (value?.testErrorSummary?.toLowerCase()
                    .includes(filterFailureInput.toLowerCase())) ||
                (value?.appErrorSummaryKeyword?.toLowerCase()
                    .includes(filterFailureInput.toLowerCase()))
          });
            return matchInvocations;
          })
        );
      };

const expectedOutput = searchException();
```

**Comment by mdazfar2** [comment, marked as answer]
@NitishKumar525 Your filtering function looks mostly correct, but there're some issues.

1. you should add the missing return some() callback.
2. you should add a null-safe check for appErrorSummaryKeyword.

Here's updated code.

const searchException = () => { return new Map( Array.from(input).filter(([key, value]) => { const matchInvocations = value.invocations.some((invocation) => invocation?.radar?.includes(filterFailureInput) );

```bash
  const matchErrorSummary =
    value?.testErrorSummary?.toLowerCase().includes(filterFailureInput.toLowerCase()) ||
    value?.appErrorSummaryKeyword?.toLowerCase()?.includes(filterFailureInput.toLowerCase());

  return matchInvocations || matchErrorSummary;
})
```
); };

**Comment by NitishKumar525** [comment]
thank you @mdazfar2 - understood what i was doing wrong. thanks for explanation. can you confirm if your final code is properly formatted in the answer.

**Comment by huuthang201** [comment]
thanks

**Comment by Tuhin092005** [comment]
The main issue is that your `.some()` callback does not return the condition. Since you're using `{}` inside the callback, you need an explicit `return`.

You can fix it like this:

```js
const searchException = () => {
  return new Map(
    Array.from(input).filter(([key, value]) => {
      const searchValue = filterFailureInput.toLowerCase();

      const matchInvocations = value.invocations?.some((invocation) =>
        invocation?.radar?.toLowerCase().includes(searchValue)
      );

      const matchTestErrorSummary =
        value?.testErrorSummary?.toLowerCase().includes(searchValue);

      const matchAppErrorSummary =
        value?.appErrorSummary?.toLowerCase().includes(searchValue);

      const matchAppErrorSummaryKeyword =
        value?.appErrorSummaryKeyword?.toLowerCase().includes(searchValue);

      return (
        matchInvocations ||
        matchTestErrorSummary ||
        matchAppErrorSummary ||
        matchAppErrorSummaryKeyword
      );
    })
  );
};

const expectedOutput = searchException();
```

For example, when `filterFailureInput` is `"50"`, `radar: "3450"` matches because it contains `"50"`, so the entry with key `0` will be included in the resulting Map.

Also, your original code checks `appErrorSummaryKeyword`, but if you want to search `appErrorSummary` too, that property needs to be included in the condition.

**Reply by mdazfar2** [reply]
Thanks, @Tuhin092005! Yes, the main issue was the missing `return` inside the `.some()` callback. Also, adding `appErrorSummary` makes the search cover all the required fields. Appreciate the additional explanation! 🙌
